File: fhir_tx/encoder.py
#
#     Copyright © 2023 Commonwealth Scientific and Industrial Research Organisation (CSIRO)
#     ABN 41 687 119 230.
#
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
#

import numpy as np
import numpy.typing as npt
import requests
from scipy.sparse import csr_matrix, hstack, lil_matrix
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.feature_extraction import DictVectorizer
from sklearn.preprocessing import OneHotEncoder

from .closure import Closure

EXPANSION_PROPERTY_PREADOPT = (
    "http://hl7.org/fhir/5.0/StructureDefinition/"
    "extension-ValueSet.expansion.contains.property"
)

IGNORED_PROPERTIES = ["parent", "child"]


class FhirTerminologyEncoder(BaseEstimator, TransformerMixin):
    """
    Encodes terminology codes into a vector representations that contain information about their
    relationships to other codes within the terminology.

    Currently this is limited to subsumption relationships as reported by the closure operation of
    the terminology server.

    :ivar codes_: The list of codes that are included in the scope, in the order that they are
        represented within the columns of the encoding.
    :type codes_: list[str]
    :ivar displays_: The list of display names for the codes that are included in the scope, in the
        order that they are represented within the columns of the encoding.
    :type displays_: list[str]
    """

    def __init__(
        self,
        scope: str,
        tx_url: str = "https://tx.ontoserver.csiro.au/fhir",
        properties: list[str] = None,
        batch_size: int = 50000,
    ):
        """
        :param scope: A FHIR ValueSet URI that defines the scope of the codes to be encoded.
        :param tx_url: A FHIR terminology server endpoint.
        :param properties: A list of properties to include in the encoding. A single value of "*"
            will include all properties.
        :param batch_size: The number of codes to send to the terminology server at a time when
            running queries.
        """
        print(f"Expanding value set: {scope}")
        coding_batches = self._expand_scope(scope, tx_url, properties, batch_size)
        print("Expansion complete")

        print("Generating one-hot encoding...", end=" ")
        ohe = OneHotEncoder(categories=[self.codes_])
        # Use a lil_matrix as an intermediate representation that enables efficient updates.
        self._encoded = lil_matrix(
            ohe.fit_transform(np.array(self.codes_).reshape(-1, 1))
        )
        print(self._encoded.shape)

        # Create an index of code -> index, which we use to update the correct row and column in the
        # matrix when making subsumption updates.
        print("Creating index...", end=" ")
        self._index = {value: index for index, value in enumerate(ohe.categories_[0])}
        print(f"{len(self._index)} items")

        print("Applying transitive closure...")
        self._apply_closure(coding_batches, tx_url)
        print(f"Subsumption encoding complete: {self._encoded.shape}")

        print("Encoding attributes...", end=" ")
        dv = DictVectorizer()
        encoded_attributes = dv.fit_transform(self.properties_)
        self._encoded = hstack([self._encoded, encoded_attributes])
        print(self._encoded.shape)

        # Create a list of feature names that includes the codes and the attributes.
        self.feature_names_ = self.codes_ + dv.feature_names_

        # Convert the final product back to a csr_matrix for efficient arithmetic operations.
        self._encoded = self._encoded.tocsr()

    def _expand_scope(self, scope, tx_url, properties, batch_size):
        """
        Get the list of all codes in the scope.
        """
        offset = 0
        total = 0
        self.codes_ = []
        self.displays_ = []
        self.properties_ = []
        coding_batches = []

        # Run an expand request with a count equal to the batch size, and iterate until we have
        # retrieved all codes.
        while offset <= total:
            params = {
                "url": scope,
                "count": batch_size,
                "offset": offset,
            }
            params = [
                ("url", scope),
                ("count", batch_size),
                ("offset", offset),
            ]
            if properties is not None:
                for p in properties:
                    params.append(("property", p))
            response = requests.get(
                f"{tx_url}/ValueSet/$expand",
                params=params,
            )
            response.raise_for_status()
            response_json = response.json()
            total = response_json["expansion"]["total"]

            print(
                f"Expanding ({len(response_json['expansion']['contains'])} items, offset {offset}, "
                f"total {total})"
            )
            codings = response_json["expansion"]["contains"]

            # Add the batch of codings to the list of batches. These same batches will be used later
            # in the closure operation.
            coding_batches.append(codings)

            # Add each code and display to a list. These are useful for later retrieval for the
            # creation of feature name dictionaries.
            self.codes_.extend([coding["code"] for coding in codings])
            self.displays_.extend(
                [
                    (coding["display"] if "display" in coding else None)
                    for coding in codings
                ]
            )
            if properties is not None:
                self.properties_.extend(
                    [properties_to_dict(coding) for coding in codings]
                )

            # The offset of the next query is the current offset plus the batch size.
            offset += batch_size

        return coding_batches

    def _apply_closure(self, coding_batches, tx_url):
        """
        Perform a closure operation on all the codes in the scope and update the encoded matrix with
        the subsumption relationships.
        """
        closure = Closure(tx_url=tx_url)
        num_batches = len(coding_batches)

        for i, batch in enumerate(coding_batches):
            print(f"Batch {i + 1} of {num_batches}, {len(batch)} items...", end=" ")
            # Get the new pairs that result from adding a batch of codes to the closure table.
            pairs = closure.update(batch)
            for pair in pairs:
                # Use the index to find the correct x and y coordinates in the matrix.
                x = self._index[pair[0]]
                y = self._index[pair[1]]
                # Update the matrix with the new subsumption relationship.
                self._encoded[x, y] = 1
            print(f"{len(pairs)} pairs added")

    def transform_column(self, X: npt.NDArray) -> csr_matrix:
        """
        Retrieve the encodings for a single column of codes.
        """
        try:
            X_indices = np.array([self._index[x] for x in X])
        except KeyError as e:
            raise ValueError(f"Encountered code not in scope: {e}")
        return self._encoded[X_indices]

    def transform(self, X: npt.NDArray[npt.NDArray]) -> csr_matrix:
        """
        Retrieve the encodings for a two-dimensional array of codes.
        """
        if not isinstance(X, np.ndarray):
            raise ValueError("X must be a Numpy array")
        if len(X.shape) != 2:
            raise ValueError("X must be a two-dimensional array")

        stacked = hstack([self.transform_column(x) for x in X])
        return stacked

    def fit(self, X, y=None):
        """
        Does nothing, but required for compatibility with scikit-learn pipelines.
        """
        return self


def properties_to_dict(coding: dict):
    try:
        result = {}
        if "property" in coding:
            for property_element in coding["property"]:
                code = property_element["code"]
                if code not in IGNORED_PROPERTIES:
                    value_key = next(k for k in coding.keys() if k.startswith("value"))
                    result[code] = property_element[value_key]
                    add_subproperties_to_dict(property_element, code, result)
        elif "extension" in coding:
            property_extensions = [
                e
                for e in coding["extension"]
                if e["url"] == EXPANSION_PROPERTY_PREADOPT
            ]
            for property_extension in property_extensions:
                code_extension = next(
                    e for e in property_extension["extension"] if e["url"] == "code"
                )
                code = code_extension["valueCode"]
                if code not in IGNORED_PROPERTIES:
                    value_extension = next(
                        (
                            e
                            for e in property_extension["extension"]
                            if e["url"] == "value"
                        ),
                        None,
                    )
                    if value_extension is not None:
                        value_key = next(
                            k for k in value_extension.keys() if k.startswith("value")
                        )
                        result[code] = value_extension[value_key]
                    add_subproperties_to_dict(property_extension, code, result)

    except StopIteration:
        return {}
    return result


def add_subproperties_to_dict(property_element: dict, code: str, result: dict):
    if "extension" in property_element:
        subproperty_extensions = [
            e for e in property_element["extension"] if e["url"] == "subproperty"
        ]
        try:
            for subproperty_extension in subproperty_extensions:
                code_extension = next(
                    e for e in subproperty_extension["extension"] if e["url"] == "code"
                )
                subcode = f"{code}.{code_extension['valueCode']}"
                if subcode not in IGNORED_PROPERTIES:
                    value_extension = next(
                        e
                        for e in subproperty_extension["extension"]
                        if e["url"] == "value"
                    )
                    value_key = next(
                        k for k in value_extension.keys() if k.startswith("value")
                    )
                    result[subcode] = value_extension[value_key]
                    add_subproperties_to_dict(subproperty_extension, subcode, result)
        except StopIteration:
            return

File: r/R/encoder.R
#
#     Copyright © 2026, Commonwealth Scientific and Industrial Research
#     Organisation (CSIRO) ABN 41 687 119 230.
#
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
#
#     Author: John Grimes
#

#' Build a FHIR terminology encoder
#'
#' Expands a FHIR ValueSet URI against a terminology server and builds a sparse
#' encoding of every code in the resulting scope. Each row of the encoding
#' carries a 1 for the code itself and a 1 for every code in the scope that
#' subsumes it, so that codes which are ontologically close share features.
#'
#' The encoder is fitted eagerly at construction: all terminology server
#' requests happen here, and `transform()` afterwards is a pure lookup.
#'
#' When `properties` are requested, the encoding is widened with one column per
#' distinct property value observed in the scope. A property value contributes
#' an indicator column named `property=value`, a numeric property value
#' contributes a single column named `property` carrying the number, and nested
#' subproperties are flattened onto their parent, as in
#' `609096000.116676008=108369006`. The property columns follow the code
#' columns, in byte order of their names.
#'
#' Progress is reported with `message()` and can be silenced with
#' `suppressMessages()`.
#'
#' @param scope A FHIR ValueSet URI defining the codes to encode, such as a
#'   SNOMED CT implicit value set built from an ECL expression - see the
#'   examples.
#' @param tx_url A FHIR terminology server endpoint. Defaults to the CSIRO
#'   Ontoserver public endpoint.
#' @param subsumption Whether to include subsumption relationships in the
#'   encoding. When `FALSE` the encoding is a plain one-hot encoding.
#' @param properties Property codes to include in the encoding, or `NULL` for
#'   none. A single value of `"*"` requests all properties.
#' @param batch_size The number of codes to send to the terminology server at a
#'   time, both when expanding and when updating the closure table.
#' @return An object of class `fhir_tx_encoder`: a list with `codes` (the codes
#'   in expansion order), `displays` (their display terms, `NA` where the server
#'   sent none), `feature_names` (the name of each column of the encoding),
#'   `encoded` (a `Matrix::dgCMatrix` with one row per code), `index` (a named
#'   integer vector mapping each code to its row and column) and `scope`.
#' @examples
#' \dontrun{
#' # Ancestors of the SNOMED CT concept "Malignant neoplastic disease".
#' encoder <- fhir_tx_encoder(
#'   scope = "http://snomed.info/sct?fhir_vs=ecl/(%3E%3E%20363346000)"
#' )
#' encoder
#'
#' # Encode two concepts.
#' transform(encoder, matrix(c("399981008", "363346000"), ncol = 1))
#' }
#' @export
fhir_tx_encoder <- function(scope,
                            tx_url = "https://tx.ontoserver.csiro.au/fhir",
                            subsumption = TRUE,
                            properties = NULL,
                            batch_size = 50000) {
  expansion <- expand_scope(scope, tx_url, properties, batch_size)
  codes <- expansion$codes
  size <- length(codes)

  # The base encoding is the identity: every code carries itself. It is built
  # from triplets rather than by mutating a sparse matrix, which would copy the
  # whole matrix on every update.
  message(sprintf("Generating one-hot encoding... (%d, %d)", size, size))
  rows <- seq_len(size)
  columns <- seq_len(size)

  # The index maps a code to its row and column, and is how a closure pair is
  # turned into a cell.
  message(sprintf("Creating index... %d items", size))
  index <- stats::setNames(seq_len(size), codes)

  if (subsumption) {
    message("Applying transitive closure...")
    cells <- closure_cells(expansion$coding_batches, tx_url, index)
    rows <- c(rows, cells$rows)
    columns <- c(columns, cells$columns)
  }

  # A cell set by more than one pair must still hold 1, not a count.
  distinct <- !duplicated(cbind(rows, columns))
  encoded <- Matrix::sparseMatrix(
    i = rows[distinct], j = columns[distinct], x = 1,
    dims = c(size, size)
  )
  if (subsumption) {
    message(sprintf(
      "Subsumption encoding complete: (%d, %d)", nrow(encoded), ncol(encoded)
    ))
  }

  # The property features are appended after the code columns, so the first n
  # columns of a widened encoding are still the codes.
  feature_names <- codes
  if (!is.null(properties)) {
    vectorized <- vectorize_properties(expansion$properties)
    encoded <- cbind(encoded, vectorized$matrix)
    feature_names <- c(feature_names, vectorized$feature_names)
    message(sprintf(
      "Encoding properties... (%d, %d)", nrow(encoded), ncol(encoded)
    ))
  }

  structure(
    list(
      codes = codes,
      displays = expansion$displays,
      feature_names = feature_names,
      encoded = encoded,
      index = index,
      scope = scope
    ),
    class = "fhir_tx_encoder"
  )
}

#' Collect the matrix cells implied by the transitive closure of a scope.
#'
#' Runs one closure update per expansion batch, so that the server is never
#' sent more codes at once than the caller asked for.
#'
#' @param coding_batches The raw codings of each expansion page.
#' @param tx_url A FHIR terminology server endpoint.
#' @param index The code to position index.
#' @return A list with `rows` and `columns` integer vectors.
#' @noRd
closure_cells <- function(coding_batches, tx_url, index) {
  name <- closure_initialize(tx_url)
  batch_count <- length(coding_batches)
  rows <- integer()
  columns <- integer()

  for (position in seq_len(batch_count)) {
    batch <- coding_batches[[position]]
    pairs <- closure_update(tx_url, name, batch)
    # The target of a pair subsumes its source, so the cell to set is
    # [row(source), col(target)].
    rows <- c(rows, unname(index[pairs$source]))
    columns <- c(columns, unname(index[pairs$target]))
    message(sprintf(
      "Batch %d of %d, %d items... %d pairs added",
      position, batch_count, length(batch), length(pairs$source)
    ))
  }

  list(rows = rows, columns = columns)
}

#' Encode a matrix of codes
#'
#' Looks up the encoding of every code in `X`. The columns of `X` are encoded
#' independently and bound side by side, so a two-column input over an encoder
#' with `f` features produces `2 * f` columns.
#'
#' @param _data An object of class `fhir_tx_encoder`.
#' @param X A two-dimensional character matrix of codes, with one row per
#'   observation and one column per code variable.
#' @param ... Unused, present for compatibility with the `transform()` generic.
#' @return A `Matrix::dgCMatrix` with `nrow(X)` rows and
#'   `ncol(X) * length(_data$feature_names)` columns.
#' @examples
#' \dontrun{
#' encoder <- fhir_tx_encoder(
#'   scope = "http://snomed.info/sct?fhir_vs=ecl/(%3E%3E%20363346000)"
#' )
#' transform(encoder, matrix(c("399981008", "363346000"), ncol = 1))
#' }
#' @export
transform.fhir_tx_encoder <- function(`_data`, X, ...) {
  if (!is.matrix(X) || !is.character(X)) {
    stop("X must be a two-dimensional character matrix", call. = FALSE)
  }
  encoded <- lapply(
    seq_len(ncol(X)),
    function(column) transform_column(`_data`, X[, column])
  )
  do.call(cbind, encoded)
}

#' Encode a single column of codes.
#'
#' @param encoder An object of class `fhir_tx_encoder`.
#' @param codes A character vector of codes.
#' @return A `Matrix::dgCMatrix` with one row per code.
#' @noRd
transform_column <- function(encoder, codes) {
  positions <- unname(encoder$index[codes])
  unknown <- is.na(positions)
  if (any(unknown)) {
    stop("Encountered code not in scope: ", codes[unknown][[1]], call. = FALSE)
  }
  encoder$encoded[positions, , drop = FALSE]
}

#' Summarise a FHIR terminology encoder
#'
#' @param x An object of class `fhir_tx_encoder`.
#' @param ... Unused, present for compatibility with the `print()` generic.
#' @return `x`, invisibly.
#' @examples
#' \dontrun{
#' encoder <- fhir_tx_encoder(
#'   scope = "http://snomed.info/sct?fhir_vs=ecl/(%3E%3E%20363346000)"
#' )
#' print(encoder)
#' }
#' @export
print.fhir_tx_encoder <- function(x, ...) {
  cat(sprintf(
    "<fhir_tx_encoder> %s - %d codes, %d features\n",
    x$scope, length(x$codes), length(x$feature_names)
  ))
  invisible(x)
}
